contracts/DefaultStrategy.sol

```diff
@@ -522,248 +443,289 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
-     */
-    function subtractFromStrategyTotalStCeloVotes(address strategy, uint256 stCeloAmount) internal {
-        defaultStrategyTotalStCeloVotes[strategy] -= stCeloAmount;
-        totalStCeloInDefaultStrategy -= stCeloAmount;
+        for (uint256 j = 0; j < groupsIndex; j++) {
+            finalGroups[j] = groups[j];
+            finalVotes[j] = votes[j];
+        }
     }
 
-    /**
-     * @notice Returns a list of group addresses with their corresponding
-     * current total votes, sorted by the number of votes, and the total number
-     * of votes in the system.
-     * @param groups The array of addresses of the groups to sort.
-     * @return The array of GroupWithVotes structs, sorted by number of votes.
-     * @return The total number of votes assigned to active groups.
-     */
-    function getSortedGroupsWithVotes(address[] memory groups)
-        internal
+    function getLesserAndGreaterOfActiveGroupsDeposit(address originalKey, uint256 newValue)
+        private
         view
-        returns (GroupWithVotes[] memory, uint256)
+        returns (address previous, address next)
     {
-        GroupWithVotes[] memory groupsWithVotes = new GroupWithVotes[](groups.length);
-        uint256 totalVotes = 0;
-        for (uint256 i = 0; i < groups.length; i++) {
-            uint256 actualCeloInGroup = account.getCeloForGroup(groups[i]);
-            uint256 stCeloInStrategy = defaultStrategyTotalStCeloVotes[groups[i]];
-            uint256 votes = Math.min(actualCeloInGroup, IManager(manager).toCelo(stCeloInStrategy));
-            totalVotes += votes;
-            groupsWithVotes[i] = GroupWithVotes(groups[i], votes);
-        }
+        uint256 loopLimit = sortingLoopLimit + 1;
+        address originalKeyPrevious;
+        previous = originalKey;
+        (, originalKeyPrevious, next) = activeGroups.get(originalKey);
 
-        sortGroupsWithVotes(groupsWithVotes);
-        return (groupsWithVotes, totalVotes);
-    }
-
-    /**
-     * @notice Sorts an array of GroupWithVotes structs based on increasing
-     * `votes` values.
-     * @param groupsWithVotes The array to sort.
-     * @dev This is an in-place insertion sort. In general in Solidity we should
-     * be careful of algorithms on arrays, especially O(n^2) ones, but here
-     * we're guaranteed to be working with a small array, its length is bounded
-     * by the maximum number of groups that can be voted for in Elections.sol.
-     */
-    function sortGroupsWithVotes(GroupWithVotes[] memory groupsWithVotes) internal pure {
-        for (uint256 i = 1; i < groupsWithVotes.length; i++) {
-            uint256 j = i;
-            while (j > 0 && groupsWithVotes[j].votes < groupsWithVotes[j - 1].votes) {
-                (groupsWithVotes[j], groupsWithVotes[j - 1]) = (
-                    groupsWithVotes[j - 1],
-                    groupsWithVotes[j]
-                );
-                j--;
+        while (next != address(0) && loopLimit-- > 1) {
```

Comment:
Unfortunately it will fail like this because of unary operator being run even when cycle is not entered.

I have rewritten it into more readable form